libnetwork/internal/l2disco/unsol_arp_linux.go

```diff
@@ -0,0 +1,80 @@
+package l2disco
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"slices"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	arpTemplate = []byte{
+		0x00, 0x01, // Hardware type
+		0x08, 0x00, // Protocol
+		0x06,       // Hardware address length
+		0x04,       // IPv4 address length
+		0x00, 0x01, // ARP request
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Sender MAC
+		0x00, 0x00, 0x00, 0x00, // Sender IP
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Target MAC (always zeros)
+		0x00, 0x00, 0x00, 0x00, // Target IP
+	}
+	bcastMAC = []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
+)
+
+type UnsolARP struct {
+	pkt []byte
+	sd  int
+	sa  *unix.SockaddrLinklayer
+}
+
+// NewUnsolARP returns a pointer to an object that can send unsolicited ARPs on
+// the interface with ifIndex, for ip and mac.
+func NewUnsolARP(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolARP, error) {
+	sd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM, int(htons(unix.ETH_P_ARP)))
```

Comment:
I'm the one telling you we need to call `CloseOnExec` below, but I now realize we should instead `|` the 3rd argument with `SOCK_CLOEXEC`.

[`man 2 socket` says](https://man7.org/linux/man-pages/man2/socket.2.html#:~:text=Since%20Linux%202.6.27%2C%20the%20type%20argument%20serves%20a%20second%20purpose%3A%20in%0A%20%20%20%20%20%20%20addition%20to%20specifying%20a%20socket%20type%2C%20it%20may%20include%20the%20bitwise%0A%20%20%20%20%20%20%20OR%20of%20any%20of%20the%20following%20values%2C%20to%20modify%20the%20behavior%20of%0A%20%20%20%20%20%20%20socket()%3A):

       Since Linux 2.6.27, the type argument serves a second purpose: in
       addition to specifying a socket type, it may include the bitwise
       OR of any of the following values, to modify the behavior of
       socket():

Reply:
Doh, I guess I've spent too long away from C! Thank you.